### PR TITLE
Accessor and Mutator for action in q_learning

### PR DESCRIPTION
**Pull request by nishantkr18**
Hey!
This should give the ability to change the actions for the agent. So with state, action, reward and terminal exposed for the agent, we can run the agent on an external env.

### CODE REVIEW

File: src/mlpack/methods/reinforcement_learning/q_learning.hpp [hunk 1]
```diff
@@ -111,6 +111,11 @@ class QLearning
   //! Get the state of the agent.
   const StateType& State() const { return state; }
 
+  //! Modify the action of the agent.
+  ActionType& Action() { return action; }
```
Comment (favre49) [inline src/mlpack/methods/reinforcement_learning/q_learning.hpp:115]: Does the setter function do anything here? Correct me if I'm wrong, but i think the action is chosen by the policy anyway, so a user setting an action wouldn't do anything?

Comment (nishantkr18) [reply src/mlpack/methods/reinforcement_learning/q_learning.hpp:115]: Hmm, u r right. I just added it thinking maybe we might use it in the future. But lets remove that for now.. Thanks :)

Comment (nishantkr18) [reply src/mlpack/methods/reinforcement_learning/q_learning.hpp:115]: BTW, I always get confused. Is it that all PRs, no matter what changes they make, should be added to HISTORY.md? Is there a threshold as such?

Comment (favre49) [reply src/mlpack/methods/reinforcement_learning/q_learning.hpp:115]: The threshold is probably different for different people, but imo its when we add new user facing things, it should be in the changelog (or big internal changes too)

**Comment by nishantkr18** [comment]
> Can you add a simple test to make sure the action is from the agent?

I'm not so sure how to do that. I guess one way is to confirm if the action returned is equal to that sampled from policy (in deterministic case).. but that would require making `policy` public in order to call `policy.Update()` from outside the class. Please correct me if I am wrong

Also, I think once our rendered env is ready and we test our agent there, then getting high rewards in the rendered environment, and seeing the agent complete the task, itself should serve as a test for action being from the agent itself, right?

**Comment by nishantkr18** [comment]
> Do you think we should add documentation for running our Q learning framework step wise, and getting the action? Currently we have only documented using it with `Episode()`.

I dont think we have used Q_learning in a stepwise manner anywhere. But, we'd be using it in the jupyter notebook for env rendering. So, we could add comments there itself. I think the purpose of the tutorial is to just enable the user to understand how to get a q_learning agent running. Please let me know what you think about that.

File: src/mlpack/methods/reinforcement_learning/q_learning.hpp (hunk 1 above)
Comment (nishantkr18) [reply src/mlpack/methods/reinforcement_learning/q_learning.hpp]: Oh, I see

**Comment by zoq** [comment]
> > Can you add a simple test to make sure the action is from the agent?
> 
> I'm not so sure how to do that. I guess one way is to confirm if the action returned is equal to that sampled from policy (in deterministic case).. but that would require making `policy` public in order to call `policy.Update()` from outside the class. Please correct me if I am wrong
> 
> Also, I think once our rendered env is ready and we test our agent there, then getting high rewards in the rendered environment, and seeing the agent complete the task, itself should serve as a test for action being from the agent itself, right?

I was thinking we could use an existing test and just check if the action size (`n_elem`) is not zero.

File: src/mlpack/tests/q_learning_test.cpp [hunk 2]
```diff
@@ -95,6 +95,8 @@ BOOST_AUTO_TEST_CASE(CartPoleWithDQN)
       break;
     }
   }
+  // to check if the action returned by the agent is not nan and is finite
```
Comment (zoq) [inline src/mlpack/tests/q_learning_test.cpp:98]:
```suggestion
  // To check if the action returned by the agent is not nan and is finite.
```

Pedantic style issue.

**Comment by zoq** [comment]
Let's see if this comes back green (except the Linux Plain build).

**Pull request merged**